src/Provider/DecoratingCacheSettingsProvider.php

```diff
@@ -228,66 +228,101 @@ private function clearIfNeeded(): void
-                if (!$lock->acquire()) {
-                    usleep(self::LOCK_RETRY_INTERVAL_MS);
-                    $this->warmup($domainNames, $settingNames);
+            $lock = $this->lockFactory->createLock(__FUNCTION__.$domainName);
+            if (!$lock->acquire()) {
+                usleep(self::LOCK_RETRY_INTERVAL_MS);
+                $this->warmupByDomainNames($domainNames);
 
-                    return;
-                }
+                return;
+            }
 
-                try {
-                    $this->warmupDomainSettings($domainName);
-                } finally {
-                    $lock->release();
-                }
-            } else {
-                foreach ($settingNames as $settingName) {
-                    if ($this->isSettingWarm($domainName, $settingName)) {
-                        continue;
-                    }
-
-                    $lock = $this->lockFactory->createLock(__FUNCTION__.$domainName.$settingName);
-                    if (!$lock->acquire()) {
-                        usleep(self::LOCK_RETRY_INTERVAL_MS);
-                        $this->warmup($domainNames, $settingNames);
-
-                        return;
-                    }
-
-                    try {
-                        $this->warmupSingleSetting($domainName, $settingName);
-                    } finally {
-                        $lock->release();
-                    }
-                }
+            try {
+                $this->warmupDomainSettings($domainName);
+            } finally {
+                $lock->release();
             }
         }
+
         $this->cache->commit();
     }
 
-    private function warmupSingleSetting(string $domainName, string $settingName): void
+    private function warmupBySettingNames(array $domainNames, array $settingNames = null): void
     {
-        $settings = $this->decoratingProvider->getSettingsByName([$domainName], [$settingName]);
-        $settingKey = $this->getSettingKey($domainName, $settingName);
+        $missingDomainNames = [];
+        $missingSettingNames = [];
 
-        if (empty($settings)) {
-            $this->storeCached($this->getCached($settingKey), null, false);
+        foreach ($domainNames as $domainName) {
+            if ($this->isDomainSettingsWarm($domainName)) {
+                continue;
+            }
 
-            return;
+            foreach ($settingNames as $settingName) {
+                if ($this->isSettingWarm($domainName, $settingName)) {
+                    continue;
+                }
+
+                $missingDomainNames[$domainName] = $domainName;
+                $missingSettingNames[$settingName] = $settingName;
+            }
+        }
+
+        if (count($missingSettingNames) > 0) {
+            $missingDomainNames = array_values($missingDomainNames);
+            $missingSettingNames = array_values($missingSettingNames);
+
+            $lockKey = implode(',', array_merge($missingDomainNames, $missingSettingNames));
```

Comment:
I don't this this lock is doing it's job. Lock is set for whole set of missing domains/settings, but items are being added to cache one by one. So if different process calls this method it might have different set (so different lock as well), but part of he same domains/settings.